src/exercise/DiscountExercise.sol

```diff
@@ -0,0 +1,138 @@
+    /// Library usage
+    using SafeTransferLib for ERC20;
+    using FixedPointMathLib for uint256;
+
+    /// Errors
+    error Exercise__SlippageTooHigh();
+    error Exercise__PastDeadline();
+
+    /// Events
+    event Exercised(address indexed sender, address indexed recipient, uint256 amount, uint256 paymentAmount);
+    event SetOracle(IOracle indexed newOracle);
+    event SetTreasury(address indexed newTreasury);
+
+    /// Immutable parameters
+
+    /// @notice The token paid by the options token holder during redemption
+    ERC20 public immutable paymentToken;
+
+    /// @notice The underlying token purchased during redemption
+    ERC20 public immutable underlyingToken;
+
+    /// Storage variables
+
+    /// @notice The oracle contract that provides the current price to purchase
+    /// the underlying token while exercising options (the strike price)
+    IOracle public oracle;
+
+    /// @notice The treasury address which receives tokens paid during redemption
+    address public treasury;
+
+    constructor(
+        OptionsToken oToken_,
+        address owner_,
+        ERC20 paymentToken_,
+        ERC20 underlyingToken_,
+        IOracle oracle_,
+        address treasury_
+    ) BaseExercise(oToken_) Owned(owner_) {
+        paymentToken = paymentToken_;
+        underlyingToken = underlyingToken_;
+        oracle = oracle_;
+        treasury = treasury_;
+
+        emit SetOracle(oracle_);
+        emit SetTreasury(treasury_);
+    }
+
+    /// External functions
+
+    /// @notice Exercises options tokens to purchase the underlying tokens.
+    /// @dev The oracle may revert if it cannot give a secure result.
+    /// @param from The user that is exercising their options tokens
+    /// @param amount The amount of options tokens to exercise
+    /// @param recipient The recipient of the purchased underlying tokens
+    /// @param params Extra parameters to be used by the exercise function
+    function exercise(address from, uint256 amount, address recipient, bytes memory params)
+        external
+        virtual
+        override
+        onlyOToken
+        returns (bytes memory data)
+    {
+        return _exercise(from, amount, recipient, params);
+    }
+
+    /// Owner functions
+
+    /// @notice Sets the oracle contract. Only callable by the owner.
+    /// @param oracle_ The new oracle contract
+    function setOracle(IOracle oracle_) external onlyOwner {
+        oracle = oracle_;
+        emit SetOracle(oracle_);
+    }
+
+    /// @notice Sets the treasury address. Only callable by the owner.
+    /// @param treasury_ The new treasury address
+    function setTreasury(address treasury_) external onlyOwner {
+        treasury = treasury_;
+        emit SetTreasury(treasury_);
+    }
+
+    /// Internal functions
+
+    function _exercise(address from, uint256 amount, address recipient, bytes memory params)
+        internal
+        virtual
+        returns (bytes memory data)
```

Comment:
maybe standardize what data you plan on returning in the future to make it easier to decode, even if you dont use it, encodepacked will just add 0s which has low gas impact

Reply:
leave it alone for now, build other exercise contracts first, than standardize across exercise contracts